lib/render.js

```diff
@@ -186,56 +191,52 @@ MapnikSource.prototype._renderMetatile = function(options, callback) {
-                        if (meta.tiles.length > 1) {
-                            renderStats.render = Math.round((Date.now() - renderStartTime) / meta.tiles.length);
-                            sliceMetatile(source, image, options, meta, renderStats, callback);
-                        } else {
-                            renderStats.render = Date.now() - renderStartTime;
-                            encodeSingleTile(source, image, options, meta, renderStats, callback);
-                        }
-                    });
-                } catch(err) {
-                    process.nextTick(function() {
-                        // Release after the .render() callback returned
-                        // to avoid mapnik errors.
-                        source._pool.release(map);
-                    });
-                    return callback(err);
-                }
-            });
-        } catch (err) {
-            return callback(err);
-        }
+    const keys = meta.tiles.map(function(tile) {
+        return options.format + ',' + tile.join(',');
     });
 
+    // acquire can throw if pool is draining
+    try {
+        this._pool.acquire((err, map) => {
+            if (err) return callback(err, keys);
+
+            // Begin at metatile boundary.
+            options.x = meta.x;
+            options.y = meta.y;
+            options.variables.zoom = options.z;
+            map.resize(meta.width, meta.height);
+            map.extent = meta.bbox;
+            try {
+                source._stats.render++;
+                const renderStats = {};
+                const renderStartTime = Date.now();
+                map.render(image, options, (err, image) => {
+                    this._pool.release(map);
+                    if (err) {
+                        return callback(err, keys);
+                    }
+                    if (meta.tiles.length > 1) {
+                        renderStats.render = Math.round((Date.now() - renderStartTime) / meta.tiles.length);
+                        sliceMetatile(source, image, options, meta, renderStats, (err, tiles) => {
+                            return callback(err, keys, tiles);
+                        });
+                    } else {
+                        renderStats.render = Date.now() - renderStartTime;
+                        encodeSingleTile(source, image, options, meta, renderStats, (err, tiles) => {
+                            return callback(err, keys, tiles);
+                        });
+                    }
+                });
+            } catch(err) {
+                this._pool.release(map);
+                return callback(err, keys);
+            }
+        });
+    } catch (err) {
+        return callback(err, keys);
+    }
+
     // Return a list of all the tile coordinates that are being rendered
     // as part of this metatile.
-    return meta.tiles.map(function(tile) {
-        return options.format + ',' + tile.join(',');
-    });
+    return keys;
```

Comment:
Is that still needed?

Reply:
Yes, it's needed because the function passed to `LockingCache` has to return the ids of the tiles generated.

Reply:
Ok, I didn't notice the `return` when calling `_renderMetatile()`. I think, we can extract:
```
const keys = meta.tiles.map(function(tile) {
   return options.format + ',' + tile.join(',');
});
```
to a method to be called from `_renderMetatile` and `LockingCache`'s generator.  Don't you mind give a try?

Reply:
I tried but since stopped because to calculate the keys you need the output from `calculateMetatile` and that was being called in `_renderMetatile`. I'll see if I can move that call into createMetaTile instead so it doesn't need to get called twice.

Reply:
Done in bb20d50.